src/main/java/seedu/address/model/room/booking/Bookings.java

```diff
@@ -3,201 +3,168 @@
-     * The Booking must exist in the list.
+     * Removes the equivalent Booking from a copy of the set.
+     * The Booking must exist in the set.
      */
-    public void remove(Booking toRemove) {
+    public Bookings remove(Booking toRemove) {
         requireNonNull(toRemove);
-        if (!internalList.remove(toRemove)) {
+        if (!sortedBookingsSet.contains(toRemove)) {
             throw new BookingNotFoundException();
         }
+        SortedSet<Booking> editedBookings = new TreeSet<>(this.sortedBookingsSet);
+        editedBookings.remove(toRemove);
+        return new Bookings(editedBookings);
     }
 
     /**
-     * Removes all expired bookings from the list.
-     */
-    public void clearExpiredBookings() {
-        internalList.removeIf(Booking::isExpired);
-    }
-
-    /**
-     * Replaces the Booking {@code target} in the list with {@code editedBooking}.
-     * {@code target} must exist in the list.
+     * Replaces the Booking {@code target} in a copy of the set with {@code editedBooking}.
+     * {@code target} must exist in the set.
      */
-    public void setBooking(Booking target, Booking editedBooking) {
+    public Bookings updateBooking(Booking target, Booking editedBooking) {
         requireAllNonNull(target, editedBooking);
 
-        int index = internalList.indexOf(target);
-        if (index == -1) {
+        if (!sortedBookingsSet.contains(target)) {
             throw new BookingNotFoundException();
         }
 
         if (!target.isSameBooking(editedBooking) && !canAcceptIfReplaceBooking(target, editedBooking)) {
             throw new OverlappingBookingException();
         }
 
-        internalList.set(index, editedBooking);
+        SortedSet<Booking> editedBookings = new TreeSet<>(this.sortedBookingsSet);
+        editedBookings.remove(target);
+        editedBookings.add(editedBooking);
+        return new Bookings(editedBookings);
     }
 
+    //=========== Boolean checkers =============================================================
 
-    public void setBookings(Bookings replacement) {
-        requireNonNull(replacement);
-        internalList.setAll(replacement.internalList);
+    /**
+     * Returns true if there are no bookings
+     */
+    public boolean isEmpty() {
+        return sortedBookingsSet.isEmpty();
     }
 
     /**
-     * Replaces the contents of this list with {@code Bookings}.
-     * {@code Bookings} must not contain duplicate Bookings.
+     * Returns true if the given booking overlaps with any existing booking in the set
      */
-    public void setBookings(List<Booking> bookings) {
-        requireAllNonNull(bookings);
-        if (!bookingsAreNotOverlapping(bookings)) {
-            throw new OverlappingBookingException();
-        }
-
-        internalList.setAll(bookings);
+    public boolean canAcceptBooking(Booking toCheck) {
+        requireNonNull(toCheck);
+        return sortedBookingsSet.stream().noneMatch(toCheck::isOverlapping);
```

Comment:
Not a major issue, but this section can actually be done in parallel!

Reply:
Also applies to other methods which require iterating through many `Room`s or `Booking`s.

Reply:
Okay noted. Same can be said for everywhere `stream` is used.